nexus/mgs-updates/src/rot_bootloader_updater.rs

```diff
@@ -6,35 +6,375 @@
+                        found: found_stage0_next_version,
+                    });
+                }
+            };
+
+            // The status is only considered ready for update if the stage0_next
+            // version found in the caboose is valid, and it matches what we
+            // found in the stage_0 caboose. Otherwise, we are waiting for an
+            // ongoing update.
+            match found_stage0_next_version {
+                FoundVersion::Version(v) => {
+                    if v == found_stage0_version {
+                        Ok(PrecheckStatus::ReadyForUpdate)
+                    } else {
+                        Ok(PrecheckStatus::WaitingForOngoingUpdate)
+                    }
+                }
+                FoundVersion::MissingVersion => {
+                    Ok(PrecheckStatus::WaitingForOngoingUpdate)
+                }
+            }
+        }
+        .boxed()
     }
 
     /// Attempts once to perform any post-update actions (e.g., reset the
     /// device)
     fn post_update<'a>(
         &'a self,
-        _log: &'a slog::Logger,
-        _mgs_clients: &'a mut MgsClients,
-        _update: &'a PendingMgsUpdate,
-    ) -> BoxFuture<'a, Result<(), GatewayClientError>> {
-        // TODO-K: To be completed in a follow up PR
-        todo!()
+        log: &'a slog::Logger,
+        mgs_clients: &'a mut MgsClients,
+        update: &'a PendingMgsUpdate,
+    ) -> BoxFuture<'a, Result<(), PostUpdateError>> {
+        async move {
+            // Before setting stage0 to the new version we want to ensure
+            // the image is good and we're not going to brick the device.
+            // We'll reset the device, causing it to check the signature.
+            // Then we'll validate that signature before we activate the
+            // new stage0.
+            debug!(
+                log,
+                "attempting to reset device to do bootloader signature check"
+            );
+            mgs_clients
+                .try_all_serially(log, move |mgs_client| async move {
+                    mgs_client
+                        .sp_component_reset(
+                            update.sp_type,
+                            update.slot_id,
+                            &SpComponent::ROT.to_string(),
+                        )
+                        .await
+                })
+                .await?;
+
+            // We now retrieve boot info from the RoT to verify the reset
+            // has completed and signature checks done.
+            debug!(
+                log,
+                "attempting to retrieve boot info to verify image validity"
+            );
+            let stage0next_error = wait_for_stage0_next_image_check(
+                log,
+                mgs_clients,
+                update.sp_type,
+                update.slot_id,
+                WAIT_FOR_BOOT_INFO_TIMEOUT,
+            )
+            .await?;
+            // If the image is not valid we bail
```

Comment:
What would it mean for the update if the image is not valid here? Or maybe: how could this happen?